app/src/main/java/com/android/wildex/ui/social/UserIndex.kt

```diff
@@ -0,0 +1,145 @@
+package com.android.wildex.ui.social
+
+import com.android.wildex.model.RepositoryProvider
+import com.android.wildex.model.social.SearchDataProvider
+import com.android.wildex.model.user.User
+import com.android.wildex.model.user.UserRepository
+import java.util.regex.Matcher
+import java.util.regex.Pattern
+
+/**
+ * Represents a result of the search, a user that corresponds to the query along with a score of how
+ * close the query is to the user
+ */
+private data class ScoredMatch(val score: Int, val string: String)
+
+/**
+ * Factor to apply to the score of a user if the query is the beginning of the user's name, surname
+ * or username
+ */
+private const val WORD_START_FACTOR = 4
+
+/**
+ * Factor to apply to the score of a user if the query is the end of the user's name, surname or
+ * username
+ */
+private const val WORD_END_FACTOR = 2
+
+/**
+ * Determines and ranks users that match a search query.
+ *
+ * @property searchDataProvider gets the users data in order to determine matches to queries
+ * @property userRepository gets the matching users in order to display them on the screen
+ */
+class UserIndex(
+    private val searchDataProvider: SearchDataProvider,
+    private val userRepository: UserRepository = RepositoryProvider.userRepository
+) {
+  /** Regex corresponding to a space to cut queries into composing words */
+  private val subQuerySeparator = Regex("\\s+")
+
+  /**
+   * Mappings from non-accented letters to all of their accented versions so that if a user types
+   * leonard -> Léonard isn't discarded as a result
+   */
+  private val expansions =
+      mapOf(
+          "c" to "[cç]",
+          "a" to "[aáàâä]",
+          "e" to "[eéèêë]",
+          "i" to "[iíìîï]",
+          "o" to "[oóòôö]",
+          "u" to "[uúùûü]")
+
+  /** Gets the users search data from the data provider */
+  private fun searchData() = searchDataProvider.getSearchData()
+
+  /**
+   * Computes all users matching to a given query but only returns at most the top N=limit ones
+   *
+   * @param query the search query to match users to
+   * @param limit the maximum number of wanted results
+   * @return the list of matching users
+   */
+  suspend fun usersMatching(query: String, limit: Int): List<User> {
+    if (searchDataProvider.dataNeedsUpdate.value) {
+      searchDataProvider.invalidateCache()
+    }
+    val searchDataToUserIds: Map<String, String> = searchData()
+
+    val patterns = subQuerySeparator.split(query).filter { it.isNotBlank() }.map { patternFor(it) }
+
+    val suggestionIds =
+        searchDataToUserIds.keys
+            .asSequence()
+            .mapNotNull { totalScore(it, patterns) }
+            .sortedByDescending { it.score }
+            .mapNotNull { searchDataToUserIds[it.string] }
+            .distinct()
+            .take(limit)
+            .toList()
```

Comment:
Part of the the exclude suggestion in the general comment


```suggestion
    val suggestionIds =
        searchDataToUserIds.keys
            .asSequence()
            .mapNotNull { totalScore(it, patterns) }
            .sortedByDescending { it.score }
            .mapNotNull { searchDataToUserIds[it.string] }
            .filter { !excludeId.contains(it) }
            .distinct()
            .take(limit)
            .toList()
```